wpcom-vip-two-factor/sms-provider.php

```diff
@@ -0,0 +1,290 @@
+		$hashed_token = get_user_meta( $user_id, self::TOKEN_META_KEY, true );
+		if ( wp_hash( $token ) !== $hashed_token ) {
+			$this->delete_token( $user_id );
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Delete the user token.
+	 *
+	 * @param int $user_id User ID.
+	 */
+	public function delete_token( $user_id ) {
+		delete_user_meta( $user_id, self::TOKEN_META_KEY );
+	}
+
+	/**
+	 * Generate and send the user token.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function generate_and_send_token( $user ) {
+		$code = $this->generate_token( $user->ID );
+		return self::send_sms( $user->phone, $code );
+	}
+
+	/**
+	 * Prints the form that prompts the user to authenticate.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function authentication_page( $user ) {
+		if ( ! $user ) {
+			return;
+		}
+
+		$this->generate_and_send_token( $user );
+		require_once( ABSPATH .  '/wp-admin/includes/template.php' );
+		?>
+		<p><?php esc_html_e( 'A verification code has been sent to the phone number associated with your account.', 'two-factor' ); ?></p>
+		<p>
+			<label for="authcode"><?php esc_html_e( 'Verification Code:', 'two-factor' ); ?></label>
+			<input type="tel" name="two-factor-sms-code" id="authcode" class="input" value="" size="20" pattern="[0-9]*" />
+		</p>
+		<?php
+		submit_button( __( 'Log In', 'two-factor' ) );
+	}
+
+	/**
+	 * Validates the users input token.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 * @return boolean
+	 */
+	public function validate_authentication( $user ) {
+		if ( ! isset( $user->ID ) || ! isset( $_REQUEST['two-factor-sms-code'] ) ) {
+			return false;
+		}
+
+		return $this->validate_token( $user->ID, $_REQUEST['two-factor-sms-code'] );
+	}
+
+	/**
+	 * Whether this Two Factor provider is configured and available for the user specified.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 * @return boolean
+	 */
+	public function is_available_for_user( $user ) {
+		return ! empty( $user->phone );
+	}
+
+	/**
+	 * Inserts markup at the end of the user profile field for this provider.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function user_options( $user ) {
+		$sms = $user->phone;
+		?>
+		<div>
+			<?php
+			echo esc_html( sprintf(
+				/* translators: %s: sms address */
+				__( 'Authentication codes will be sent to %s.', 'two-factor' ),
+				$sms
+			) );
+			?>
+		</div>
+		<?php
+	}
+
+	/**
+	 * Send an SMS message to the specified numbers using the Twilio API
+	 *
+	 * @param int|array Nnumber(s) to send an SMS to. Accepts formatted and unformatted US numbers, e.g. +14155551212, (415) 555-1212 or 415-555-1212.
+	 * @param string Message to send. Longer than 140 chars will be split.
+	 *
+	 * @link http://www.twilio.com/docs/api/rest/sending-sms
+	 */
+	static function send_sms( $to, $message, $country_code = '+1' ) {
```

Comment:
`send_sms`, `send_single_sms_via_rest `, `get_rest_url`, and `split_words` should probably be in their own class.